File: 05 09 25/validsudoku.php
<?php
class Solution {
    function solveSudoku(&$board) {
        $this->solve($board);
    }

    private function solve(&$board) {
        for ($i = 0; $i < 9; $i++) {
            for ($j = 0; $j < 9; $j++) {
                if ($board[$i][$j] === '.') {
                    for ($c = '1'; $c <= '9'; $c++) {
                        if ($this->isValid($board, $i, $j, $c)) {
                            $board[$i][$j] = $c;
                            if ($this->solve($board)) return true;
                            $board[$i][$j] = '.'; // backtrack
                        }
                    }
                    return false;
                }
            }
        }
        return true;
    }

    private function isValid($board, $row, $col, $c) {
        for ($i = 0; $i < 9; $i++) {
            if ($board[$row][$i] === $c || $board[$i][$col] === $c) return false;
            $boxRow = 3 * intdiv($row, 3) + intdiv($i, 3);
            $boxCol = 3 * intdiv($col, 3) + $i % 3;
            if ($board[$boxRow][$boxCol] === $c) return false;
        }
        return true;
    }
}

File: September/5/validsudoku.rb
class Solution
  def solve_sudoku(board)
    solve(board)
  end

  private

  def solve(board)
    (0..8).each do |i|
      (0..8).each do |j|
        if board[i][j] == '.'
          ('1'..'9').each do |c|
            if is_valid?(board, i, j, c)
              board[i][j] = c
              return true if solve(board)
              board[i][j] = '.'  # backtrack
            end
          end
          return false
        end
      end
    end
    true
  end

  def is_valid?(board, row, col, c)
    (0..8).each do |i|
      return false if board[row][i] == c || board[i][col] == c
      box_row = 3 * (row / 3) + i / 3
      box_col = 3 * (col / 3) + i % 3
      return false if board[box_row][box_col] == c
    end
    true
  end
end
